core/src/main/java/io/grpc/internal/DnsNameResolver.java

```diff
@@ -130,18 +123,16 @@ public final synchronized void refresh() {
       public void run() {
         InetAddress[] inetAddrs;
         Listener savedListener;
-        synchronized (DnsNameResolver.this) {
-          // If this task is started by refresh(), there might already be a scheduled task.
-          if (resolutionTask != null) {
-            resolutionTask.cancel(false);
-            resolutionTask = null;
-          }
-          if (shutdown) {
-            return;
-          }
-          savedListener = listener;
-          resolving = true;
+        // If this task is started by refresh(), there might already be a scheduled task.
```

Comment:
Hmm, you are right. ChannelExecutor needs to be exposed to NameResolver so that tasks ran in an executor can be serialized with it. I will put this PR on hold until I have time to improve it.